File: leetcode/0040-combination-sum-ii/solution.py
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        path = []

        def backtrack(start: int, remaining: int) -> None:
            if remaining == 0:
                result.append(path[:])
                return

            for i in range(start, len(candidates)):
                if candidates[i] > remaining:
                    break  # sorted, so nothing further can work either
                if i > start and candidates[i] == candidates[i - 1]:
                    continue  # skip duplicate picks at this recursion level

                path.append(candidates[i])
                backtrack(i + 1, remaining - candidates[i])
                path.pop()

        backtrack(0, target)
        return result

File: leetcode/0040-combination-sum-ii/Solution.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    private int[] candidates;
    private List<List<Integer>> result;
    private List<Integer> path;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        this.candidates = candidates;
        this.result = new ArrayList<>();
        this.path = new ArrayList<>();

        backtrack(0, target);
        return result;
    }

    private void backtrack(int start, int remaining) {
        if (remaining == 0) {
            result.add(new ArrayList<>(path));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (candidates[i] > remaining) {
                break;
            }
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }

            path.add(candidates[i]);
            backtrack(i + 1, remaining - candidates[i]);
            path.remove(path.size() - 1);
        }
    }
}
